File: Lokesh_Saana/77-combinations.py
# https://leetcode.com/problems/combinations/

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []

        def backtrack(start, comb):
            if len(comb) == k:
                res.append(comb.copy())
                return

            if start > n:
                return

            for i in range(start, n+1):
                comb.append(i)
                backtrack(i+1, comb)
                comb.pop()

        backtrack(1, [])
        return res

File: Ayush-Gupta/77. Combinations.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        int i=0;
        vector<vector<int>> v;
        vector<int> p(k,0);
        while(i>=0) {
            p[i]++;
            if(p[i]>n) i--;
            else if(i==k-1) v.push_back(p);
            else {
                i++;
                p[i]=p[i-1];
            }
        }
        return v;
    }
};